airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py

```diff
@@ -132,46 +131,27 @@ def filepath_iterator() -> Iterator[str]:
         :yield: url filepath to use in StorageFile()
         """
 
-    def pattern_matched_filepath_iterator(self, filepaths: Iterable[str]) -> Iterator[str]:
+    def pattern_matched_filepath_iterator(self, file_infos: Iterable[FileInfo]) -> Iterator[FileInfo]:
         """
-        iterates through iterable filepaths and yields only those filepaths that match user-provided path patterns
+        iterates through iterable file_infos and yields only those file_infos that match user-provided path patterns
 
-        :param filepaths: filepath_iterator(), this is a param rather than method reference in order to unit test this
+        :param file_infos: filepath_iterator(), this is a param rather than method reference in order to unit test this
         :yield: url filepath to use in StorageFile(), if matching on user-provided path patterns
         """
-        for filepath in filepaths:
-            if globmatch(filepath, self._path_pattern, flags=GLOBSTAR | SPLIT):
-                yield filepath
+        for file_info in file_infos:
+            if globmatch(file_info.key, self._path_pattern, flags=GLOBSTAR | SPLIT):
+                yield file_info
 
     @lru_cache(maxsize=None)
-    def get_time_ordered_filepaths(self) -> Iterable[Tuple[datetime, str]]:
+    def get_time_ordered_file_infos(self) -> List[FileInfo]:
         """
         Iterates through pattern_matched_filepath_iterator(), acquiring last_modified property of each file to return in time ascending order.
         Uses concurrent.futures to thread this asynchronously in order to improve performance when there are many files (network I/O)
         Caches results after first run of method to avoid repeating network calls as this is used more than once
 
         :return: list in time-ascending order
         """
-
-        def get_storagefile_with_lastmod(filepath: str) -> Tuple[datetime, str]:
-            fc = self.storagefile_class(filepath, self._provider)
-            return (fc.last_modified, filepath)
-
-        storagefiles = []
-        # use concurrent future threads to parallelise grabbing last_modified from all the files
-        # TODO: don't hardcode max_workers like this
-        with concurrent.futures.ThreadPoolExecutor(max_workers=64) as executor:
-
-            filepath_gen = self.pattern_matched_filepath_iterator(self.filepath_iterator())
-
-            futures = [executor.submit(get_storagefile_with_lastmod, fp) for fp in filepath_gen]
-
-            for future in concurrent.futures.as_completed(futures):
-                # this will failfast on any errors
-                storagefiles.append(future.result())
-
-        # The array storagefiles contain tuples of (last_modified, filepath), so sort by last_modified
-        return sorted(storagefiles, key=itemgetter(0))
+        return sorted(self.pattern_matched_filepath_iterator(self.filepath_iterator()), key=lambda file_info: file_info.last_modified)
```

Comment:
This has potential to reduce performance with many files but probably is a very small % compared to the extra actual data read runtime so not totally against this.

Updating docstring to reflect this changed functionality would be 👌

Reply:
updated